src/nodes/strings.js

```diff
@@ -86,12 +91,39 @@ module.exports = {
       return preferSingleQuotes ? "''" : "\"\"";
     }
 
+    const isSingleQuote = tstring_beg === "'";
+    const isDoubleQuote = tstring_beg === "\"";
+    const preferDoubleQuotes = !preferSingleQuotes;
+
     // Determine the quote to use. If we prefer single quotes and there are no
     // embedded expressions and there aren't any single quotes in the string
     // already, we can safely switch to single quotes.
-    let quote = "\"";
-    if (preferSingleQuotes && string.body.every(isSingleQuotable)) {
+    let quote = undefined;
+
+    if (isSingleQuote && preferSingleQuotes) {
       quote = "\'";
+    } else if (isDoubleQuote && preferDoubleQuotes) {
+      quote = "\""
+    } else if (isSingleQuote && preferDoubleQuotes) {
+      if (string.body.every(isDoubleQuotable)) {
+        quote = "\"";
+      } else {
+        quote = "\'";
+      }
+    } else if (isDoubleQuote && preferSingleQuotes) {
+      if (string.body.every(isSingleQuotable)) {
+        quote = "\'";
+      } else {
+        quote = "\"";
+      }
+    } else if (isDoubleQuote && preferDoubleQuotes) {
+      if (string.body.every(isDoubleQuotable)) {
+        quote = "\""
+      } else {
+        quote = "\'"
+      }
+    } else {
+      throw new Error("Could not correctly format string literal");
```

Comment:
I decided to go for explicit rather than code golfing this. Additionally, this error path shouldn't happen.